### ISSUE TITLE
random intervol repeating keys, not mechanical

### ISSUE THREAD

**Issue by baos780** [pwr/Solaar#264]
Having an odd issue where at random moments my '5' key starts repeating.  
Have replaced the k800 keyboard with another brand new k800 keyboard, replaced the usb cord, replaced the dongle.  Also unplugged the mouse(Logitech G602, not a solaar mouse/conflict?) while it was happening, it continued.  Shut off keyboard while it was happening, continued.  Pulled keyboard dongle while it was happening and it stopped.  It's also possible it is simply a logitech defect and not the issue with solaar.  
$ uname -a
Linux darren-pc 4.4.0-18-generic #34-Ubuntu SMP Wed Apr 6 14:01:02 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux

Solaar 0.9.2

**Comment by Lekensteyn** [comment]
Could it be a late April 1st joke (that happened before)?

**Comment by baos780** [comment]
It's been happening for a few months,  Usually happens once every 24-48 hours.  It stops when I actually type the '5' key however this is the second k800 keyboard.  If it's an april fools joke, it's in the source code.

**Comment by Lekensteyn** [comment]
(Not a joke from Solaar, but maybe a friend.)

It could be a power management issue of the USB controller. Does it also happen with other ports? Does it happen on other machines as well? What if you leave the machine unused for 48 hours?

**Comment by baos780** [comment]
I suspect the power issues and bought a cheap usb power monitor, but would a repeating 5 and no other key be too specific to be a power issue?  From my reading, some gigabyte motherboards are plagued with bad power design.

**Comment by Lekensteyn** [comment]
Hmm, some minutes ago I also noticed 5s in my Firefox location bar (laptop and K800 keyboard as well). No idea if I accidentally dropped something on my keyboard or if something else is going on.

I only remember one other occassion where a key would continuously be pressed, don't know if that was 5 or something else. Since you seem to be able to reproduce it, can you obtain a USB capture using Wireshark? You only need the "dumpcap" binary included with wireshark-common on Ubuntu. Identify the bus number `X` in lsusb and capture with `dumpcap -i usbmonX -w usb.pcapng`. Be careful not to type passwords while capturing. If you captured the issue, you can also cut the capture slightly before the moment of occurrence.

**Comment by baos780** [comment]
Sounds good. I'll probably have something in 48 hours.

**Comment by baos780** [comment]
My dumpcap is usbmon4, however I have a logitech g602 mouse which operates on a different frequency with a different receiver which is not supported by solaar.  However usbmon4 seems to capture the mouse, or at least I watch the packet count go up when I move the mouse(again not through a unifying receiver).

**Comment by Lekensteyn** [comment]
Multiple USB devices can reside on the same bus, so try moving around devices until you have isolated the bus to one device. (Check with `lsusb -t` for a tree).

**Comment by Lekensteyn** [comment]
I can now also "reproduce" the issue. Left my keyboard on over the night and woke up with typing 5s at the login screen (SDDM, KDE Plasma). This happened yesterday at least twice (once at the login screen, once while developing).

Other possibly helpful information:

 - M525 mouse is turned off.
 - K800 keyboard was turned on, but on low battery.
 - Linux 4.5.0-1-ARCH

You can use this command to extract the USB communication from the pcap if you are interested:

    tshark -T fields -e frame.time -e usb.capdata -e usb.data_fragment -Y "usb.capdata or usb.data_fragment" -r usb1.pcapng

(computer -> receiver is on the right (`usb.data_fragment`), receiver -> computer on the left (`usb.capdata`).

Ignoring register reads, I don't see anything extraordinary from the receiver->computer. I'll need more time later to analyse this.

**Comment by baos780** [comment]
Got the cap, not sure how to make sense of it.  
[usb.pcapng.gz](https://github.com/pwr/Solaar/files/228172/usb.pcapng.gz)

**Comment by baos780** [comment]
The bug has been solved.  https://bugs.launchpad.net/ubuntu/+source/linux/+bug/1579190  
[quote]
Kamal Mostafa (kamalmostafa) wrote on 2016-06-16: 	#31

Tim, its essentially the same reason that we saw the problem manifest in some Wily and Xenial kernel versions, then seem to disappear in the next kernel version, only to return again in some future version:

The bug, which definitely existed all the way back through Precise, was that the driver could access the wrong data (an out of bounds array access) in some circumstances. The theory is that for any each individual kernel build, that particular wrong data it accesses might end up containing 'the 5 key' (making this bug manifest) or might end up containing a 'null' (resulting in no obvious manifestation for that kernel build). Its plausible that in Precise, the kernel builds just always ended up with a 'null' there.

The fix patch ([mainline] 50220de HID: core: prevent out-of-bound readings) prevents the driver from accessing that wrong data at all, which is certainly the correct behavior whether or not the 55555 problem was seen to manifest.

[/quote]
